File: codes/java/src/polymorphism/TypeManip.java
package polymorphism;

import java.lang.reflect.*;

public class TypeManip {
    public static void testMembers(Object obj) {
        try {
            Method learnM = obj.getClass().getMethod("learn");
            learnM.invoke(obj, new Object[]{});
        }
        catch (Throwable t) {
            System.out.println("Method learn does not exist!");
        }

        try {
            Field nbrF = obj.getClass().getField("nbr");
            System.out.println("Field nbr = " + nbrF.get(obj));
        }
        catch (Throwable t) {
            System.out.println("Field nbr does not exist!");
        }
    }
    public static void main(String[] args) {
        Person pe = new Person();
        Person st = new Student();
        Person pr = new Professor();

        System.out.println("==========================");
        System.out.println("Object type");
        System.out.println("==========================");
        System.out.println("Type of pe: " + pe.getClass().getName());
        System.out.println("Type of st: " + st.getClass().getName());
        System.out.println("Type of pr: " + pr.getClass().getName());

        System.out.println("==========================");
        System.out.println("Instance of");
        System.out.println("==========================");
        System.out.println("pe instanceof Person: " + (pe instanceof Person));
        System.out.println("st instanceof Person: " + (st instanceof Person));
        System.out.println("st instanceof Student: " + (st instanceof Student));
        System.out.println("st instanceof Professor: " + (st instanceof Professor));
        System.out.println("pr instanceof Person: " + (pr instanceof Person));
        System.out.println("pr instanceof Student: " + (pr instanceof Student));
        System.out.println("pr instanceof Professor: " + (pr instanceof Professor));
        System.out.println("pr instanceof Worker: " + (pr instanceof Worker));

        System.out.println("==========================");
        System.out.println("Member existance");
        System.out.println("==========================");
        testMembers(pe);
        testMembers(st);
        testMembers(pr);

        System.out.println("==========================");
        System.out.println("Casting");
        System.out.println("==========================");
        Student st2 = (Student) st;
        st2.learn();
        Professor pr2 = (Professor) pr;
        pr2.teach();
        pr2.work();
        ((Worker) pr).work();
    }
}

class Person {
    public void talk(){
        System.out.println("I am talking");
    }
}

class Student extends Person {
    public void learn(){
        System.out.println("I am learning");
    }
}

interface Worker {
    default public void work(){
        System.out.println("I am working");
    }
}

class Professor extends Person implements Worker {
    public int nbr = 5;
    public void teach(){
        System.out.println("I am teaching");
    }
}

File: codes/python/polymorphism/type_manip.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

class Person(object):
    def talk(self):
        print("I am a talking")

class Student(Person):
    def learn(self):
        print("I am learning")

class Professor(Person):
    def __init__(self):
        self.nbr = 5
    def teach(self):
        print("I am teaching")

class PhdStudent(Student, Professor): pass

pe = Person()
st = Student()
pr = Professor()
phd = PhdStudent()

print("==========================")
print("Object type")
print("==========================")
print("type(pe): " + type(pe).__name__)
print("st.__class__: " + st.__class__.__name__)
print("type(pr): " + type(pr).__name__)
print("type(phd): " + type(phd).__name__)
print("type(25): " + type(25).__name__)

print("==========================")
print("Instance of")
print("==========================")

print ("isinstance(pe, Person) " + str(isinstance(pe, Person)))
print ("isinstance(st, Person) " + str(isinstance(st, Person)))
print ("isinstance(25, Person) " + str(isinstance(25, Person)))
print ("isinstance(st, Student) " + str(isinstance(st, Student)))
print ("isinstance(st, Professor) " + str(isinstance(st, Professor)))
print ("isinstance(pr, Professor) " + str(isinstance(pr, Professor)))
print ("isinstance(phd, Professor) " + str(isinstance(phd, Professor)))
print ("isinstance(phd, Student) " + str(isinstance(phd, Student)))

print("==========================")
print("Member existance")
print("==========================")
import sys
def testMembers(msg, obj):
    sys.stdout.write(msg + ": ")
    learn = getattr(obj, "learn", None)
    if learn != None and callable(learn):
        learn()
    else:
        print("Method learn does not exist!")
    sys.stdout.write(msg + ": ")
    if hasattr(obj, "nbr"):
        print("Field nbr = " + str(obj.nbr))
    else:
        print("Field nbr does not exist!")

testMembers("Person", pe)
testMembers("Student", st)
testMembers("Professor", pr)
testMembers("PhdStudent", phd)
testMembers("a number", 25)
